src/param_env.md

```diff
@@ -0,0 +1,17 @@
+# Parameter Environment
+
+When working with associated constants or generic types it is often relevant to
+have more information about the `Self` or generic parameters. Trait bounds and
+similar information is encoded in the `ParamEnv`. Often this is not enough
+information to obtain things like the type's `Layout`, but you can do all kinds
+of other checks on it (e.g. whether a type implements `Copy`) or you can
+evaluate an associated constant whose value does not depend on anything from the
+parameter environment.
+
+Although you can obtain a valid `ParamEnv` for any item via
+`tcx.param_env(def_id)`, this `ParamEnv` can be too generic for your use case.
+Using the `ParamEnv` from the surrounding context can allow you to evaluate more
+things.
```

Comment:
I don't understand this sentence. It seems like you should use the param-env that is correct, and no other. =) Maybe the question is more about determining what is correct? Can you give a specific example of what you had in mind here?

Reply:
`promoted_const_def_id` doesn't exist though?

Reply:
hmm... maybe it was associated constants and not promoteds.

I'm definitely using the `param_env` of the surrounding function when evaluating constants, not just when calling `Instance::resolve`.

Reply:
The rule is that you use the `ParamEnv` for the body you're monomorphizing. So a function referring to a constant or anything else, will use whatever is in the MIR, substituted with the `Substs` and the `ParamEnv` (if needed) *for the function*.

That is, if you got the MIR for a `def_id`, typesystem things you get *from that MIR* are valid wrt `tcx.param_env(def_id).subst(tcx, substs)`, after applying `.subst(tcx, substs)`.